File: python/yacht/yacht.py
from collections import Counter

YACHT = 0
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = 7
FOUR_OF_A_KIND = 8
LITTLE_STRAIGHT = 9
BIG_STRAIGHT = 10
CHOICE = 11


def yacht(dice):
    return 50 if len(set(dice)) == 1 else 0


def sum_of_xs(x):
    return lambda dice: sum(number for number in dice if number == x)


def full_house(dice):
    counter = Counter(dice)

    return sum(dice) if set(counter.values()) == {3, 2} else 0


def four_of_a_kind(dice):
    counter = Counter(dice)
    number, count = counter.most_common(1)[0]

    return 4 * number if count >= 4 else 0


def little_straight(dice):
    return 30 if set(dice) == {1, 2, 3, 4, 5} else 0


def big_straight(dice):
    return 30 if set(dice) == {2, 3, 4, 5, 6} else 0


def choice(dice):
    return sum(dice)


categories = [
    yacht,
    sum_of_xs(1),
    sum_of_xs(2),
    sum_of_xs(3),
    sum_of_xs(4),
    sum_of_xs(5),
    sum_of_xs(6),
    full_house,
    four_of_a_kind,
    little_straight,
    big_straight,
    choice,
]


def score(dice, category):
    return categories[category](dice)

File: haskell/yacht/src/Yacht.hs
module Yacht
  ( yacht
  , Category(..)
  ) where

import           Data.List                      ( sort )

data Category = Ones
              | Twos
              | Threes
              | Fours
              | Fives
              | Sixes
              | FullHouse
              | FourOfAKind
              | LittleStraight
              | BigStraight
              | Choice
              | Yacht

yacht :: Category -> [Int] -> Int
yacht Ones      = sum . filter (== 1)
yacht Twos      = sum . filter (== 2)
yacht Threes    = sum . filter (== 3)
yacht Fours     = sum . filter (== 4)
yacht Fives     = sum . filter (== 5)
yacht Sixes     = sum . filter (== 6)
yacht FullHouse = fullHouse . sort
 where
  fullHouse [a, b, c, d, e]
    | (a == c && c /= d && d == e) || (a == b && b /= c && c == e)
    = a + b + c + d + e
    | otherwise
    = 0
  fullHouse _ = 0
yacht FourOfAKind = fourOfAKind . sort
 where
  fourOfAKind [a, b, c, d, e] | a == d    = a + b + c + d
                              | b == e    = b + c + d + e
                              | otherwise = 0
  fourOfAKind _ = 0
yacht LittleStraight = littleStraight . sort
 where
  littleStraight [1, 2, 3, 4, 5] = 30
  littleStraight _               = 0
yacht BigStraight = bigStraight . sort
 where
  bigStraight [2, 3, 4, 5, 6] = 30
  bigStraight _               = 0
yacht Choice = sum
yacht Yacht  = yacht' . sort
 where
  yacht' [a, _, _, _, e] | a == e    = 50
                         | otherwise = 0
  yacht' _ = 0
